### PR TITLE
Add Tabular provider

### PR DESCRIPTION
**Pull request by Fokko**
---
First version of the Tabular hook :)

**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/tabular/hooks/tabular.py
Comment (kbendick) [inline airflow/providers/tabular/hooks/tabular.py]: Nit: `dbt Cloud` -> `Tabular`.

File: airflow/providers/tabular/hooks/tabular.py
Comment (kbendick) [inline airflow/providers/tabular/hooks/tabular.py]: Nit: I would leave the note about being used with an AWS managed identity off until such a time, if ever, that there's an example of that.

The first sentence alone seems sufficient for that.

File: airflow/utils/db.py
Comment (kbendick) [inline airflow/utils/db.py]: In the form widgets we have `extra__tabular__baseUrl`. Is this an additional user enterable field that we should add to the form widget as well?

File: docs/apache-airflow-providers-tabular/connections.rst
Comment (kbendick) [inline docs/apache-airflow-providers-tabular/connections.rst]: Nit: Is this supposed to say `Trino` connection or `Tabular` connection? I don't believe there's anything Trino specific about these.

**Review by kbendick — COMMENTED** [review]
Thanks @Fokko! Left some comments with questions about various things.

File: airflow/providers/tabular/hooks/tabular.py
Comment (Fokko) [reply airflow/providers/tabular/hooks/tabular.py]: 🙈

File: airflow/providers/tabular/hooks/tabular.py
Comment (Fokko) [reply airflow/providers/tabular/hooks/tabular.py]: Yes, I agree, thanks!

File: docs/apache-airflow-providers-tabular/connections.rst
Comment (Fokko) [reply docs/apache-airflow-providers-tabular/connections.rst]: That should be Tabular indeed, Thanks!

File: airflow/utils/db.py
Comment (Fokko) [reply airflow/utils/db.py]: Good catch! I switched that around a couple of times :)

File: .github/CODEOWNERS
Comment (turbaszek) [inline .github/CODEOWNERS:62]: 🚀

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (turbaszek) [inline airflow/providers/tabular/example_dags/example_tabular.py]: Would you mind following `with DAG(...)` convention that seems to prevail in example dags?

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (turbaszek) [inline airflow/providers/tabular/example_dags/example_tabular.py]: Isn't this code invoked every time a DAG is parsed?

Comment (kaxil) [reply airflow/providers/tabular/example_dags/example_tabular.py]: Indeed

File: tests/providers/tabular/hooks/test_tabular.py
Comment (turbaszek) [inline tests/providers/tabular/hooks/test_tabular.py]:
```suggestion
class TestTabularHook(unittest.TestCase):
```

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (Fokko) [reply airflow/providers/tabular/example_dags/example_tabular.py]: Good call! Let me update this 👍🏻

File: tests/providers/tabular/hooks/test_tabular.py
Comment (Fokko) [reply tests/providers/tabular/hooks/test_tabular.py]: Thanks!

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (Fokko) [reply airflow/providers/tabular/example_dags/example_tabular.py]: Hmm, good catch. Let me think about a solution

Comment (turbaszek) [reply airflow/providers/tabular/example_dags/example_tabular.py]: Maybe `user_defined_macros` that will be lazily evaluated?

File: airflow/models/dag.py
Comment (eladkal) [inline airflow/models/dag.py]: Is this change related to this PR?

**Review by eladkal — COMMENTED** [review]
There is an ongoing discussion in mailing list about if we should add more providers
https://lists.apache.org/thread/nvfc75kj2w1tywvvkw8ho5wkx1dcvgrn

though I would say this is a unique case as a PMC is the codeowner :)

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (eladkal) [reply airflow/providers/tabular/example_dags/example_tabular.py]: We have macros for connections ` {{  conn.test_conn.login }}`  etc...
but there is also another problem - rendering it means that the token is exposed in the render tab of the UI.

File: airflow/models/dag.py
Comment (Fokko) [reply airflow/models/dag.py]: That's actually incorrect, and shouldn't have been committed :) Thanks!

**Comment by Fokko** [comment]
Thanks for pointing me to the thread @eladkal

I actually started on a 3rd party provider, but setting it up within Airflow itself makes it so much easier because of Breeze :) @kbendick and I will maintain this one.

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (Fokko) [reply airflow/providers/tabular/example_dags/example_tabular.py]: Thanks @eladkal for the suggestion. I've changed the code to make use of the connection macro. 

I'm not seeing the actual token in the TI details:
image

I do see it in the Rendered template:
image

But templating it inside of bash isn't the way to do it. It will be passed using an environment variable, which is nicely obfuscated.

Also, I'm seeing the token because I'm explicitly echoing it:
image

It probably will be plugged into the `env_vars` of the SparkSubmit or SparkSQL operator: https://airflow.apache.org/docs/apache-airflow-providers-apache-spark/stable/_api/airflow/providers/apache/spark/operators/spark_submit/index.html

File: airflow/providers/tabular/provider.yaml
Comment (eladkal) [inline airflow/providers/tabular/provider.yaml]: Is this needed if the provider is already compatible only with >=2.2 ?

**Review by kaxil — COMMENTED** [review]
This will probably be in the same boat as https://github.com/apache/airflow/pull/22659 where we are waiting for what comes out of https://lists.apache.org/thread/nvfc75kj2w1tywvvkw8ho5wkx1dcvgrn

File: airflow/providers/tabular/provider.yaml
Comment (Fokko) [reply airflow/providers/tabular/provider.yaml]: Great catch!

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (ashb) [reply airflow/providers/tabular/example_dags/example_tabular.py]: > We have macros for connections ` {{  conn.test_conn.login }}` etc... but there is also another problem - rendering it means that the token is exposed in the render tab of the UI.

This is a bug - and it should be redacted there too.

Fix https://github.com/apache/airflow/pull/22754

**Comment by kaxil** [comment]
> This will probably be in the same boat as #22659 where we are waiting for what comes out of https://lists.apache.org/thread/nvfc75kj2w1tywvvkw8ho5wkx1dcvgrn

Welp, looks like @eladkal had already commented about it :)

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (kbendick) [inline airflow/providers/tabular/example_dags/example_tabular.py]: Nit: Possibly we make this named something more unique, even just `tabular_example`? This would be similar to the DBT example dag. Most provider example DAGs seemed to be named the same as the file itself: https://github.com/apache/airflow/blob/df84c4ad42b1bbaf352d14973f1f867c89dc0a3b/airflow/providers/dbt/cloud/example_dags/example_dbt_cloud.py#L33-L39

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (josh-fell) [inline airflow/providers/tabular/example_dags/example_tabular.py]: A while ago there was some cleanup done of example DAGs' `default_args` use (e.g. #16872 and others) to remove some less relevant args and move the `default_args` declaration directly to the `DAG()` object. It would be nice to keep the consistency.

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (josh-fell) [inline airflow/providers/tabular/example_dags/example_tabular.py]: Can you add `catchup=False`? This was another example-DAG cleanup effort (yeah, there were several) so beginner users who might use the example DAG won't have headaches with lots of DAG Runs being launched if the `start_date` doesn't change and haven't read up on `catchup`.

File: airflow/providers/tabular/hooks/tabular.py
Comment (josh-fell) [inline airflow/providers/tabular/hooks/tabular.py]: Rather than creating a new extra field, WDYT about relabeling `host` instead?

File: airflow/providers/tabular/hooks/tabular.py
Comment (josh-fell) [inline airflow/providers/tabular/hooks/tabular.py]: Not a strong opinion, but maybe drop the "Tabular" prefix given that the connection type will be set to "Tabular"?

**Comment by potiuk** [comment]
I think best to add them in AIP-47 way

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (Fokko) [reply airflow/providers/tabular/example_dags/example_tabular.py]: I like that 👍🏻

Comment (Fokko) [reply airflow/providers/tabular/example_dags/example_tabular.py]: Good one, I've added it

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (Fokko) [reply airflow/providers/tabular/example_dags/example_tabular.py]: Missed that memo. I just inlined the `default_args` 👍🏻

File: airflow/providers/tabular/hooks/tabular.py
Comment (Fokko) [reply airflow/providers/tabular/hooks/tabular.py]: Great suggestion, just fixed that

File: airflow/providers/tabular/hooks/tabular.py
Comment (Fokko) [reply airflow/providers/tabular/hooks/tabular.py]: I also don't have a super-strong opinion on it, I've dropped the tabular prefix :)

File: airflow/utils/db.py
Comment (josh-fell) [inline airflow/utils/db.py]: Since the base URL is now `host` under the hood, can you update the default connection and the connection doc when you get a chance?

**Comment by turbaszek** [comment]
Do we plan to push this PR forward?

**Comment by eladkal** [comment]
> Do we plan to push this PR forward?

We have yet to reach a conclusion about the approch for new providers
https://lists.apache.org/thread/nvfc75kj2w1tywvvkw8ho5wkx1dcvgrn
We should reach some conclusion as it currently blocks adding new providers.

File: airflow/utils/db.py
Comment (Fokko) [reply airflow/utils/db.py]: Thanks @josh-fell I forgot about this one! Much appreciated

**Comment by Fokko** [comment]
@turbaszek I would love to get this in :) I've just addressed the latest comment by @josh-fell and rebased. 🤞🏻 for a green CI.

@eladkal How about getting this in? I'm happy to migrate the code later on once we get to a conclusion.

**Comment by potiuk** [comment]
Hey @fokko - almost green :). You seem to have some extra (?) stuff in command hash file (`breeze regenerate-command-images` should help). 

Can I ask for still a little patience please? I am about to summarize a discussion on providers and propose some approach there :)

**Comment by potiuk** [comment]
> Would be great to get this in at some point, looking forward to the summary

Coming :)

**Comment by Fokko** [comment]
@potiuk I saw your mail: https://lists.apache.org/thread/6ngq79df7op541gfwntspdtsvzlv1cr6

It makes sense and I see the current issues. That being said, would it be possible to get this in? Both @kbendick and I will make sure that the operator will be maintained. One thing on the operator, the Tabular one will be quite small. I can also see that we release Apache Iceberg operators in the future as well, but that will probably also be just a wrapper around the Python pyiceberg library.

**Comment by potiuk** [comment]
> @potiuk I saw your mail: https://lists.apache.org/thread/6ngq79df7op541gfwntspdtsvzlv1cr6
> 
> It makes sense and I see the current issues. That being said, would it be possible to get this in? Both @kbendick and I will make sure that the operator will be maintained. One thing on the operator, the Tabular one will be quite small. I can also see that we release Apache Iceberg operators in the future as well, but that will probably also be just a wrapper around the Python pyiceberg library.

I believe so :). I wanted just to get some community feedback, but I think having just a commitment from someone who has an interest (be it direct stakeholder or say a company that implements a lot of integrations with the product) - even verbally and confirmed in an issue should be more than enough. Feel free to comment in the discussion I started too :).

**Comment by potiuk** [comment]
Hey @Fokko, It took a bit long but I think we got to a consensus regarding new providers, and we also propose some kind of mix-governance aproach, where (among the others) the stakeholders for the future providers (which we are going to technically split to separate repositories soon) should take a bit more responsibility for maintenance:

https://github.com/apache/airflow/pull/24680

If that does not scare you away, and you still want to add the provider to Airflow community providers, feel free to rebase the PR. I will also ask you (hopefully it will be merged soon) to rebase it after we merge #24672 - we are going to change the way how we keep depdencies for providers in order to prepare them to separate to different repository.

Let us know what you want to do, either close the PR or rebase it and lead it to completion after #24672 is merged.

**Comment by Fokko** [comment]
Hey @potiuk, of course I'm still interested in getting this in! I'm working on rebasing the PR, but I'm also seeing some issues with the recent changes. Let me dig into it and get back to you

File: airflow/providers/tabular/CHANGELOG.rst
Comment (eladkal) [inline airflow/providers/tabular/CHANGELOG.rst]: Should be 1.0.0 isn't it?

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (eladkal) [inline airflow/providers/tabular/example_dags/example_tabular.py]: new example dags should be written according to AIP-47 
the example dag should be in `system/providers/tabular` path
see https://github.com/apache/airflow/pull/24641 for reference

File: airflow/providers/tabular/provider.yaml
Comment (eladkal) [inline airflow/providers/tabular/provider.yaml]: same version question as changelog

File: airflow/providers/tabular/example_dags/example_tabular.py
Comment (Fokko) [reply airflow/providers/tabular/example_dags/example_tabular.py]: All green now :)

File: docs/apache-airflow-providers-tabular/index.rst
Comment (josh-fell) [inline docs/apache-airflow-providers-tabular/index.rst:63]:
```suggestion
Release: 1.0.0
```

File: docs/apache-airflow-providers-tabular/index.rst
Comment (josh-fell) [inline docs/apache-airflow-providers-tabular/index.rst:74]:
```suggestion
You can install this package on top of an existing Airflow 2.2+ installation via
```
The minimum version for providers was bumped to 2.2.0 recently 🙂

File: tests/providers/tabular/hooks/test_tabular.py
Comment (josh-fell) [inline tests/providers/tabular/hooks/test_tabular.py:25]: For net-new unit tests we should prefer `pytest` over `unittest`.

File: docs/apache-airflow-providers-tabular/connections.rst
Comment (josh-fell) [inline docs/apache-airflow-providers-tabular/connections.rst:1]: I _think_ we typically want to have a URI example for connection in the doc. Is that right @mik-laj?

**Review by josh-fell — APPROVED** [review]
Just some small things but LGTM.

**Review by potiuk — APPROVED** [review]
Looks cool!

File: docs/apache-airflow-providers-tabular/index.rst
Comment (potiuk) [reply docs/apache-airflow-providers-tabular/index.rst:63]: Ah.. was to quick :).. @Fokko - mind for a small follow-up?

Comment (Fokko) [reply docs/apache-airflow-providers-tabular/index.rst:63]: Sure thing!

File: docs/apache-airflow-providers-tabular/index.rst
Comment (Fokko) [reply docs/apache-airflow-providers-tabular/index.rst:74]: Missed that one, I'll do a follow-up! 👍🏻

File: tests/providers/tabular/hooks/test_tabular.py
Comment (Fokko) [reply tests/providers/tabular/hooks/test_tabular.py:25]: I actually prefer pytest as well. Let me follow up on this one

**Comment by kbendick** [comment]
Thank you @Fokko!

**Pull request merged**
